### PR TITLE
Adding EventSource/EventCounters to Sockets

### PR DESCRIPTION
**Pull request by BrennanConroy**
Part of #349 
First time Eventing, lets see how it goes

@anurse

### CODE REVIEW

File: src/Microsoft.AspNetCore.Sockets/Internal/SocketEventSource.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Sockets.Internal
+{
+    [EventSource(Name = "Microsoft-AspNetCore-Sockets")]
+    internal class SocketEventSource : EventSource
+    {
+        public static readonly SocketEventSource Log = new SocketEventSource();
+
+        private readonly EventCounter _connectionsStartedCounter;
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets/Internal/SocketEventSource.cs]: Counter is probably redundant

File: src/Microsoft.AspNetCore.Sockets/ConnectionManager.cs [hunk 2]
```diff
@@ -135,6 +138,7 @@ public void Scan()
                     // Once the decision has been made to dispose we don't check the status again
                     if (status == DefaultConnectionContext.ConnectionStatus.Inactive && (DateTimeOffset.UtcNow - lastSeenUtc).TotalSeconds > 5)
                     {
+                        SocketEventSource.Log.ConnectionTimedOut(c.Value.ConnectionId);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets/ConnectionManager.cs]: Where is the log for this?

File: src/Microsoft.AspNetCore.Sockets/ConnectionManager.cs [hunk 3]
```diff
@@ -66,16 +66,19 @@ public DefaultConnectionContext CreateConnection()
             var applicationSide = ChannelConnection.Create<byte[]>(transportToApplication, applicationToTransport);
 
             var connection = new DefaultConnectionContext(id, applicationSide, transportSide);
+            var connectionTimer = SocketEventSource.Log.ConnectionStart(id);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets/ConnectionManager.cs]: Where is the ILogger.Log for this?

Comment (BrennanConroy) [reply src/Microsoft.AspNetCore.Sockets/ConnectionManager.cs]: On line 60, I'll move this to be next to it.

File: src/Microsoft.AspNetCore.Sockets/DefaultConnectionContext.cs [hunk 4]
```diff
@@ -22,6 +23,7 @@ public class DefaultConnectionContext : ConnectionContext,
         // This tcs exists so that multiple calls to DisposeAsync all wait asynchronously
         // on the same task
         private TaskCompletionSource<object> _disposeTcs = new TaskCompletionSource<object>();
+        internal ValueStopwatch _connectionTimer;
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets/DefaultConnectionContext.cs]: Use a property. We generally avoid setting fields from outside the class. Keep it `internal` though.

File: src/Microsoft.AspNetCore.Sockets/Internal/SocketEventSource.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+                if (IsEnabled(EventLevel.Informational, EventKeywords.None))
+                {
+                    ConnectionStop(connectionId);
+                }
+            }
+        }
+
+        [Event(eventId: 1, Level = EventLevel.Informational, Message = "Started connection '{0}'.")]
+        public ValueStopwatch ConnectionStart(string connectionId)
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets/Internal/SocketEventSource.cs]: Any other values we could add to the payload? We don't know the transport yet, correct? Maybe include the request's TraceIdentifier if we have it?

File: src/Microsoft.AspNetCore.Sockets/ConnectionManager.cs (hunk 2 above)
Comment (anurse) [reply src/Microsoft.AspNetCore.Sockets/ConnectionManager.cs]: Yeah, we aren't doing the "bundle the event source call in the logger extension method" thing, but in general, it should be abnormal to write an event and **not** write a log message. So if there isn't a log message for this, add one as well :).

File: src/Microsoft.AspNetCore.Sockets/Internal/SocketEventSource.cs (hunk 5 above)
Comment (BrennanConroy) [reply src/Microsoft.AspNetCore.Sockets/Internal/SocketEventSource.cs]: We don't have any info at this point

**Pull request merged**